### ISSUE TITLE
Standardize Field Configuration

### ISSUE THREAD

**Issue by BeksOmega** [google/blockly#2722]
**Is your feature request related to a problem? Please describe.**

@RoboErikG  mentioned [here ](https://github.com/google/blockly/pull/2654#discussion_r307908212)that he would like to move to a new system for configuring field customizations.

I think this is a good idea, but I have some questions so I thought I'd move it to an issue.

**Describe the solution you'd like**

> I think I'd rather start using an opt_config parameter on the constructor with a map of the configuration options. The keys should be the same as the JSON config which will also let you use the same code for both.

**Describe alternatives you've considered**

@RoboErikG 

**Open questions**
Should this change only to apply to future fields? Or should the currently available field constructors change to support this?

Should the opt_config accept tooltips and other parameters that apply to all fields?

Should there be a hook so you can change the options at runtime (e.g. setOptions(config))?

Should the currently available hooks (e.g. setSpellcheck) be removed, deprecated, or left?

**Comment by RoboErikG** [comment]
What makes the most sense to me (and @rachel-fenichel may overrule me):

- Use this pattern for future fields that have non-trivial configuration. We can discuss what counts as non-trivial on a field by field basis.
- For existing fields that have 2+ config parameters adding an opt_config to the constructor makes sense, but is low priority.
- Existing configuration methods should be left. Slightly fewer lines of code in block inits doesn't really justify the churn of asking developers to update.
- I'm slightly against adding setOptions. Configuring fields outside of init adds a lot of complexity. Do you have a good use case in mind?

**Comment by rachel-fenichel** [comment]
+1 to @RoboErikG

**Comment by BeksOmega** [comment]
Ok I have a revised proposal! Some things are added and some things are changed, so tell me what you think of it.

Important: I have highlighted things that I think need discussion with italics.

### Proposal: Field Configuration Paradigm

Field configuration is not consistent. There are four places where field configuration is possible: JSON definition, JavaScript constructor, JavaScript constants, and JavaScript hooks.

#### Definitions and Constructors

The JSON definition and the JavaScript constructor should be equivalent in terms of function. Anything you can configure with one you should be able to configure with the other. They should both cover all configurations exposed by constants and hooks.

#### Constants

JavaScript constants modify all field instances of a given type and cannot be changed while Blockly is running. They do not need to cover all configurations available in the definition or constructor. They should only be available when there is a logical default value for a configuration, and it would be reasonable for an outside developer to set this property globally.

* A `FieldTextInput.SPELLCHECK` property would be a good constant because an outside developer could reasonably want TextInput fields to spellcheck by default.
* A `FieldLabel.CLASS` property would be a bad constant because it doesn't make sense to apply a class globally when we already (mostly) provide css.

#### Hooks

JavaScript hooks modify individual fields, and can be changed while Blockly is running. They do not need to cover all configurations available in the definition or constructor, and they do not need to cover all constants. They should only be available for configurations that can be changed while Blockly is running without adding a lot of complexity.

* `FieldTextInput.setSpellcheck` is a good hook because it is a relatively unimportant visual change that doesn't affect the value.
* `FieldAngle.setWrap` would be a bad hook because the wrap does affect the value.

All hooks should return "this" (the field) to support method chaining.

### tl:dr

Constants are globals; hooks work while Blockly's running; defs and constructors cover all constants and hooks + extras.

### (Some) Discussion Points

> [Definitions and constructors] should both cover all configurations exposed by constants and hooks.

Does anyone forsee problems with this, or have opinions about this?

> All hooks should return "this" (the field) to support method chaining.

I think this would be a good addition, and it would work backwards compatibly (I don't think there are any config methods that return). But I know there was disagreement about it on [the PR](https://github.com/google/blockly/pull/2654).

**Comment by RoboErikG** [comment]
Regarding hooks returning 'this' I think we decided against it. Chaining configuration methods is typically used when there's some form of finalization that applies all the changes you've made (like an apply() or a build() method). Because the configuration change would immediately trigger a redraw in most cases I'd rather have it not return this and be clear that each operation is an atomic change.

**Comment by BeksOmega** [comment]
I agree that not adding the method chaining would make it clearer that each change is atomic, but I'm not sure that that's necessary.

For example jQuery allows method chaining for its configuration methods:
```
$('#my-div')
  .css('background', 'blue')
  .height(100)
  .fadeIn(200);
```
Plus Blockly does as well, chaining already exists in block initialization:
```
this.appendDummyInput()
        .appendField(new Blockly.FieldLabelSerializable("first text"), "TEXT1")
        .appendField(new Blockly.FieldColour("#ff0000"), "COLOUR")
        .appendField(new Blockly.FieldLabelSerializable("second text"), "TEXT2");
```
and colour field configuration:
```
var field = new Blockly.FieldColour("#ff0000");
        .setColours(myColourArray, myTitlesArray)
        .setColumns(3);
```
I think it's definitely more readable, rather than more confusing. But that's just my two cents, I'll drop it now hehe.

I forgot to mention in my previous post that once I get confirmation I'll write up a hit list of things that need changing. So now it has officially been mentioned :D

**Comment by rachel-fenichel** [comment]
I think most constants should not control the behaviour of all fields of that type.  In the spellcheck case, for instance, I can see having some text inputs that need to be spellchecked and others that don't, on the same workspace.  Where we do have constants that apply to all fields of a type, like the ones for angle fields, I think it's an accident and/or mistake.

**Comment by BeksOmega** [comment]
> In the spellcheck case, for instance, I can see having some text inputs that need to be spellchecked and others that don't, on the same workspace.

I absolutely agree, and we do currently support setting the spellcheck at the instance level, just not at a field-type level.

I think maybe a better way to phrase what I was suggesting would be to say: "Constants would be a way to change the defaults of a field type. Then those defaults could be overriden on a per-instance basis using the JSON definition or JS constructor".

Let me reframe this with a couple of options.

Defaults (at the field-type level) can be manipulated by:
1) Changing the default value of an instance property. E.g. set `Blockly.FieldTextInput.prototype.spellcheck_ = true`
2) Changing the value of a constant property. E.g set
`Blockly.FieldTextInput.SPELLCHECK = true`

Then there are options for documentation:
A) Don't document how to manipulate the defaults of a field type.
B) Do document how to manipulate the defaults of a field type.

I'm cool with any combination of the options. I'll try to include backwards compat details in my hitlist.

**Comment by RoboErikG** [comment]
I think defaults need to be decided on a field by field basis.

A lot of fields it doesn't make sense or is very rare to need to update the default globally (disabling spellcheck for all text fields probably isn't a common use case), so documenting what the default is and that it can be set on the instance or in the constructor's options is enough.

For fields that have a strong case to change the default for all instances (like FieldAngle's picker options), then using constant properties that are documented makes sense to me.

**Comment by BeksOmega** [comment]
### Hit List
This comment details what field customization should look like at the end of this process (given my current understanding of your input).

This includes all customization options currently detailed in the built-in fields documentation, but I realize you may not want all of these to be official APIs. So be sure to shout if there's somethin' you don't like : )

#### Abstract Field
This is meant to facilitate setting global configurations like a tooltip, using the JS constructor of a specific field.
- [ ] JS constructor **changed** to function(opt_value, opt_validator, opt_config)
  - opt_config will accept {tooltip}
- [ ] an `@protected` doConfiguration_ function should be **added**
  - This function will be overriden by child fields to handle their specific configuration.
  - It should only be called by the base class (inside the constructor), and never by the child classes or developers.
  - On the base class it will handle configurations available to all fields. Currently this is just {tooltip}


#### Angle Fields
- [ ] JS constructor **changed** to function(opt_value, opt_validator, opt_config)
  - opt_config will accept {mode, clockwise, offset, wrap, round}
- [ ] JSON definition **changed** to accept {angle, mode, clockwise, offset, wrap, round}
- [X] JavaScript constants already exist and **remain unchanged** {CLOCKWISE, OFFSET, WRAP, ROUND}
- No JavaScript hooks will be provided.

#### Checkbox Fields
- [ ] JS constructor **changed** to function(opt_value, opt_validator, opt_config)
  - opt_config will accept {checkChar}
- [ ] JSON definition **changed** to accept {checked, checkChar}
- [X] A JavaScript constant already exists and will **remain unchanged** {CHECK_CHAR}
- [ ] setCheckChar function **added**

#### Colour Fields
- [ ] JS Constructor **changed** to function(opt_value, opt_validator, opt_config)
  - opt_config will accept {colourOptions, colourTitles, columns}
- [X] JSON definition is correct and will **remain unchanged**  {colour, colourOptions, colourTitles, columns}
- [X] JavaScript constants already exist and will **remain unchanged** {COLOURS, TITLES, COLUMNS}
- [X] JavaScript hooks already exist and will **remain unchanged**
  - setColours(coloursArray, titlesArray)
  - setColumns

#### Date Fields
No customizations available at this time.

#### Dropdown Fields
If you want to keep the percentage height property, I believe it should be changed to be a percentage of the workspace, rather than a percentage of the window.
- [ ] JS Constructor **changed** to function(menuGenerator, opt_validator, opt_config)
  - opt_config will accept {arrowChar, maxPercentHeight}
- [ ] JSON definition **changed** to accept {arrowChar, maxPercentHeight}
- [X] JavaScript constants already exist {ARROW_CHAR, MAX_MENU_HEIGHT_VH}
- [ ] setArrowChar function **added**

#### Image Fields
- [ ] JS constructor **changed** to function(src, width, height, opt_onClick, opt_config)
  - opt_config will accept {opt_alt, opt_flipRtl}
  - Note that if people attempt to create an image field this way using an old version of Blockly it will not work. The [domToWorkspace](https://github.com/google/blockly/blob/de39d5f2d056f0f380dbbfb8f86cfa82ea449f55/core/xml.js#L363) problem.
- [X] JSON definition correct and will **remain unchanged** {src, width, height, alt, flipRtl}
- No JavaScript constants will be provided.
- Current hooks consist of:
  - setValue for the src.
  - setAlt for the alt.
  - No hooks for setting the width, height.
  - [X] Pending further discussion JS hooks will **remain unchanged**

#### Label Fields (Serializable and Non-Serializable)
- [ ] JS constructor **changed** to function(opt_value, opt_config)
  - opt_config will accept {class}
- [X] JSON definition correct and will **remain unchanged** {text, class}
- No JavaScript constants will be provided.
- [ ] setClass function **added**

#### Number Fields
- [ ] JS constructor **changed** to function(opt_value, opt_validator, opt_config)
  - opt_config will accept {min, max, precision}
  - Note that if people attempt to create a number field this way using an old version of Blockly it will not work. The [domToWorkspace](https://github.com/google/blockly/blob/de39d5f2d056f0f380dbbfb8f86cfa82ea449f55/core/xml.js#L363) problem.
- [X] JSON definition is correct and will **remain unchanged** {value, min, max, precision}
- No JavaScript constants will be provided.
- [X] A JavaScript hook already exists and will **remain unchanged**
  - setConstraints

#### Text Input Fields
- [ ] JS constructor **changed** to function(opt_value, opt_validator, opt_config)
  - opt_config will accept {spellcheck}
- [X] JSON definition is correct and will **remain unchanged** {text, spellcheck}
- No JavaScript constants will be provided.
- [X] A JavaScript hook already exists and will **remain unchanged**
  - setSpellcheck

#### Variable Fields
No customizations available at this time (besides typing, if that counts).

**Comment by BeksOmega** [comment]
This is Beka requesting [go/no go](https://www.youtube.com/watch?v=BHIo6qwJarI) for modifying the abstract field class.

**Comment by RoboErikG** [comment]
Note on implementation: Instance properties with JS constant defaults should be null unless specified with a getter that returns the property | DEFAULT. That way changing the default at runtime will cause all unspecified field instances to also update.

Image Fields: It's worth adding a typeof check to the parameters and falling back to the old parameter order if the optional params are "string, function, boolean)".

Number fields: Ditto on typeof checks.

Otherwise, LGTM.

**Comment by BeksOmega** [comment]
This is Beka requesting go/no go for angle fields:

> * JS constructor changed to function(opt_value, opt_validator, opt_config)
>   * opt_config will accept {mode, clockwise, offset, wrap, round}
> * JSON definition changed to accept {angle, mode, clockwise, offset, wrap, round}
> * JavaScript constants already exist and remain unchanged {CLOCKWISE, OFFSET, WRAP, ROUND}
> * No JavaScript hooks will be provided.

**Comment by samelhusseini** [comment]
Thanks for the hard work @BeksOmega! Looks like we just have angle fields left!

**Comment by BeksOmega** [comment]
Are angle fields go?

**Comment by samelhusseini** [comment]
Yep! Thanks

**Comment by BeksOmega** [comment]
Ok, last thing field I need feedback on is dropdown fields:
> - [ ] **Dropdown Fields**
If you want to keep the percentage height property, I believe it should be changed to be a percentage of the workspace, rather than a percentage of the window.
>  - [ ] JS Constructor **changed** to function(menuGenerator, opt_validator, opt_config)
    - opt_config will accept {arrowChar, maxPercentHeight}
>  - [ ] JSON definition **changed** to accept {arrowChar, maxPercentHeight}
>  - [X] JavaScript constants already exist {ARROW_CHAR, MAX_MENU_HEIGHT_VH}
>  - [ ] setArrowChar function **added**

Importantly, How do you want the maxPercentHeight handled?
A) Remove it from the public API (i.e. undocument it).
B) Keep it as is, and add config opts.
C) Change it to be a percentage of the workspace height, and add config opts.
D) Somethin' else.

@samelhusseini or @RoboErikG

**Comment by samelhusseini** [comment]
On the one hand I think it's pretty cheap to add it as a config option so why not, but I wouldn't want to clutter options with all these little tweaks developers can make. I think it's unlikely this is going to be an option developers use on a field by field basis, leaving it as is makes the most sense to me and developers can override that constant if they want. 

As for whether it should be percentage of workspace vs document height, I say also keep as is, as widget divs / dropdown divs extend past the workspace, and hence a document height percentage makes more sense for "more" consistent results.

Do you think maybe arrow char can also stay as a static option for all fields? I just don't see developers changing these. I guess this falls back to what the purpose of this opt_config is, is it to surface field options on an instance level or is it to surface all options to JSON configuration?

**Comment by BeksOmega** [comment]
Every point you make sounds to me.

I will proceed with simply adding a opt_config to support tooltips. (^-^)b

**Comment by samelhusseini** [comment]
Sounds great, thank you!

**Comment by samelhusseini** [comment]
Thanks for all your help @BeksOmega!
